doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,422 @@
+catalog states capable of returning only the state we need for bootstrapping.
+
+All current usages of `mz_adapter::catalog::storage::Connection` will be updated to
+use `Box<dyn DurableCatalogState>`. The `environmentd` binary will initialize
+the `Box<dyn DurableCatalogState>` using the methods above depending on command line arguments.
+
+### Stash implementation
+
+A stash based implementation will be fairly trivial to implement. This interface is already
+implemented using the stash, we will just need to change the name of some methods.
+
+### Persist Implementation
+
+This implementation is based
+on [Aljoscha's skunk works project](https://github.com/aljoscha/materialize/tree/skunk-log-for-coord-state).
+The entire state will be stored in a single persist shard as raw bytes with a tag to differentiate
+between object types. The existing protobuf infrastructure will be used to serialize and deserialize
+objects. The implementing struct will maintain a persist write handle, an upper, and
+an in-memory cache of all the objects. Certain append only object types, like audit
+logs and storage usage, will not be cached in memory since they are only written to and not read.
+NB: The storage events are read once at start-time and never read again.
+
+#### Initialization
+
+- `init_persist_catalog_state`:
+    1. Creates a new persist write handle.
+    2. Read the shard upper into memory.
+    3. Store the boot timestamp in memory.
+- `is_initialized`: Checks if upper is equal to the minimum timestamp.
+- `open`:
+    - If the catalog is not initialized:
+        1. Use the write handle to compare and append a batch of initial catalog objects.
+        2. Update the in memory upper to the new upper.
+    - If the catalog is initialized:
+        1. Open a persist read handle.
+        2. Read a snapshot of the catalog state into memory.
+        3. Run migrations in memory (including incrementing the epoch).
+        4. Use the write handle to compare and append a batch of migration changes into persist.
+        5. Update the in memory upper to the new upper.
+- `check_open`:
+    - If the catalog is not initialized then as long as we're the leader return `Ok`.
+    - If the catalog is initialized:
+        1. Open a persist read handle.
+        2. Read a snapshot of the catalog state into memory.
+        3. Run migrations in memory.
+            - None of the objects will be cached in memory.
+- `open_read_only`:
+    - If the catalog is not initialized then fail.
+    - If the catalog is initialized:
+        1. Open a persist read handle.
+        2. Read a snapshot of the catalog state into memory.
+- `boot_ts`: Return boot timestamp from memory.
+- `epoch`: This will be stored the same as any other catalog object so this can be treated as a
+  normal [read](#reads).
+
+#### Reads
+
+All reads (except for the deploy generation) will perform the follow steps:
+
+1. Confirm leadership.
```

Comment:
Why do reads need to confirm leadership? What happens if the confirm fails? What happens if there is a race condition where the confirm succeeds but another envd becomes the leader immediately after?

Reply:
> Why do reads need to confirm leadership?

They might not need to confirm leadership just to do a read. I'm not aware of a specific correctness or other reason that leadership confirmation needs to be done. If we do confirm leadership, it might prevent us from doing the work to compute a write later on but...that just doesn't matter. An unchecked read isn't wrong: the data are correct as of some time. Hmm.

> What happens if the confirm fails?

We've been fenced out and will halt.

> What happens if there is a race condition where the confirm succeeds but another envd becomes the leader immediately after?

If this is only a read then the current envd will continue on until it does another leadership check. If the read data is used to compute a write then the write will fail because the write will also do a leadership check.

Reply:
> I'm not aware of a specific correctness or other reason that leadership confirmation needs to be done.

We need to check that we're still the leader _at some point during the execution of a read query_ in order to maintain strict serializability. But it's possible that that the check we do at a higher level than the catalog (e.g., selecting a timestamp for the query) is sufficient.

Reply:
Agree. I'm not aware of a reason we need to during catalog reads. I'm not sure if read queries have a logical dependency on catalog read access having done a leadership check; I'm pretty sure they don't.

Reply:
@jkosh44 sent me the following message:

>  I don't have access to GH but I couldn't help but try and answer your Q on my design doc. Leadership checks are needed for strict serializable reads.
> This design doc talks about the process and the race condition you asked about: https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20220516_transactional_consistency.md#reads
> Feel free to share any of that info on GH.

Reply:
Ah! Happy to have been wrong about this and the reason is written down.